### ISSUE TITLE
Make work on Functors

### ISSUE THREAD

**Issue by oxinabox** [JuliaDiff/ChainRulesTestUtils.jl#117]
right now all the tests check they are not funning on functors.

With the new API in https://github.com/JuliaDiff/ChainRulesTestUtils.jl/pull/116
we can have `f ⟂ ḟ` and `f ⟂ f̄``, which would be fine API-wise.

**Comment by piever** [comment]
Current work around (discussed on Slack) is to cheat a bit and define an intermediate function, that is

```julia
(f::F)(x) = compute(f, x)

function compute(f::F, x)
    # The actual callable struct implementation
end

function ChainRulesCore.rrule(::typeof(compute), f::F, x)
    # Compute value and pullback function
end

# Now testing works fine
test_rrule(compute, f, x)
```

**Comment by rick2047** [comment]
Why can't we facilitate this work around in the code itself? Work around seems to be simple enough that instead of checking for a functor we can wrap it ourselves?

**Comment by oxinabox** [comment]
The reason this issue is still open is that it hasn't been done yet.
Because noone has gotten round to making a PR.
Feel encouraged to open one, contributions are welcome. :-)

We don't need the work around in the code itself. when we can fix it properly.
Which is mostly just making sure things are fed to FiniteDifferences.jl correctly and removing the check.
Although, I am not 100% sure that that doesn't actually look a lot like the workaround. It might, it might not.

**Comment by rick2047** [comment]
if you provide me a few pointers on what is meant by "fed to FiniteDifferences.jl correctly", I can open a PR. Maybe some documentation?

**Comment by oxinabox** [comment]
Working out exactly what it is to fed to FiniteDIfferences.jl correctly, is kinda the hard part of the PR.



The actual methods we hit in FiniteDifferences.jl are these ones:
https://github.com/JuliaDiff/FiniteDifferences.jl/blob/c31952549df3607c4809d141c533d83d0015e23a/src/grad.jl#L49-L76
Which use `to_vec` to deal with types that are not vectors.
That already should support functors just fine
https://github.com/JuliaDiff/FiniteDifferences.jl/blob/c31952549df3607c4809d141c533d83d0015e23a/src/to_vec.jl#L26-L41
but the finite differences method are not prepared to accept functors as their `f` arg.
Which does mean  something like https://github.com/JuliaDiff/ChainRulesTestUtils.jl/issues/117#issuecomment-780758808
is wanted.

Probably  adapting the `_wrap_function` in https://github.com/JuliaDiff/ChainRulesTestUtils.jl/blob/master/src/finite_difference_calls.jl to do that is right.
That file is  kinda gross as is (which means making it more gross is less of a problem).
That whole thing (and a lot of the FiniteDifferences file) will one day be cleaned up more extensively by https://github.com/JuliaDiff/FiniteDifferences.jl/issues/97
but that is a much bigger project.
And having this work even if a bit ugly is useful for that to inform that future FiniteDifferences.jl rewrite.

The other thing that is need is to workout what (if anything) need to be done to the `make_jvp_calls`
https://github.com/JuliaDiff/ChainRulesTestUtils.jl/blob/8b5e9067cec52cf97da5cceec9bc040460c5cccd/src/finite_difference_calls.jl#L18
Mayube that should will change from:
`_make_jvp_call(fdm, f, y, xs, ẋs, ignores)`
to `_make_jvp_call(fdm, y, xs, ẋs, ignores)` with `f` just being the first element of `xs` and its derivative the first of `ẋs` and similar it having a entry in `ignores`.
similar for `_make_j′vp_call`.
and those changes would need to be propegated upwards to the places that call those.
Basically so that most of the code in this package doesn't treat `f` any different from any of its arguments.

**Comment by rick2047** [comment]
Ok I gave it my best but I can't figure it out. Thanks for the explanation though. Maybe good for the next person.
